JProxy update

mkitti opened the pull request:
I have updated the `JProxy` code from @zot from PR #91 . Tests are now passing and there are no conflicts.

zot commented:
Thanks a ton for doing this!

zot commented on src/proxy.jl:
```diff
@@ -0,0 +1,1494 @@
+
+location(source) = replace(string(source.file), r"^.*/([^/]*)$" => s"\1") * ":" * string(source.line) * ": "
+
+macro verbose(args...)
```
yes, @verbose is just for debugging the project itself and should eventually be removed or commented out. I left some of the logging commented out so I could uncomment it at need (and I did have to uncomment some of them sometimes :) )

zot commented on test/runtests.jl:
```diff
@@ -4,6 +4,7 @@ using JavaCall
 import Dates
 using Base.GC: gc
 
+const pxyptr = JavaCall.pxyptr
```
Correct, I had used that in a println in an earlier version of the test code and I forgot to remove it

mkitti replied:
These tests will be moved to a separate sub-package.

zot commented on src/JavaCall.jl:
```diff
@@ -4,7 +4,7 @@ export JavaObject, JavaMetaClass,
        @jimport, jcall, jfield, isnull,
        getname, getclass, listmethods, getreturntype, getparametertypes, classforname,
-       narrow
+       narrow, JProxy, @class, interfacehas, staticproxy
```
Maybe "JavaObject", because you use it like a Java object?

aviks replied:
Yes, that is the most relevant name, the only trouble is that we already have a JavaObject type in JavaCall, so we'll have figure out a transition. However, I think the macro should be called `@java`.

mkitti commented on src/convert.jl:
```diff
@@ -264,14 +280,16 @@ function convert(::Type{@jimport(java.util.List)}, x::Vector, V::Type{JavaObject
-          jstr.ptr, buf)
+          jstr, buf)
     return s
 end
```
@aviks Could we incorporate the changes to this file convert.jl into JavaCall proper?

aviks replied:
Sure. Is the current code likely a bug?

mkitti replied:
I think it was just quoted out of context. What I meant is that most of the changes to convert.jl seem acceptable to merge. It will have to be updated for the new JNI module though.

mkitti commented on src/core.jl:
```diff
@@ -16,8 +17,15 @@ const jdouble = Cdouble
 jprimitive = Union{jboolean, jchar, jshort, jfloat, jdouble, jint, jlong}
 
-struct JavaMetaClass{T}
+mutable struct JavaMetaClass{T}
```
Why is it necessary to change this to a mutable struct? I haven't look very hard for the answer, but it might save time if someone could explain it.

mkitti replied:
Might it make sense to have mutable and unmutable versions of JavaMetaClass?

zot replied:
I'll look around -- I'm pretty sure I had a good reason for that. But maybe not :)

zot replied:
Because Java can garbage collect classes and the finalizer on line 84 nulls out ptr for safety. If we can guarantee that deleteref will only ever be called from finalize, it can stay immutable I think.

mkitti commented on src/jvm.jl:
```diff
@@ -200,6 +201,7 @@ function init(opts)
     jvm = unsafe_load(pjvm)
     global jvmfunc = unsafe_load(jvm.JNIInvokeInterface_)
     global jnifunc = unsafe_load(jnienv.JNINativeInterface_) #The JNI Function table
+    initProxy()
```
We should revert the changes to this file for now.

zot replied:
Looks like init_current_vm() needs initProxy() as well

mkitti commented on test/Test.java:
```diff
@@ -83,4 +85,47 @@ public String innerString() {
+
+  public String testArrayArgs(Object[] i) {
+    return "java.lang.Object[]";
+  }
```
The changes to Test.java and the binaries should be reverted for now and moved into a separate sub-package.

mkitti commented on src/convert.jl:
```diff
@@ -1,19 +1,34 @@
+convert(::AbstractString, str::Type{JString}) = unsafe_string(str)
 convert(::Type{JString}, str::AbstractString) = JString(str)
 convert(::Type{JObject}, str::AbstractString) = convert(JObject, JString(str))
+convert(::Type{JavaObject{Symbol("java.lang.Double")}}, n::Real) = jnew(Symbol("java.lang.Double"), (jdouble,), Float64(n))
+convert(::Type{JavaObject{Symbol("java.lang.Float")}}, n::Real) = jnew(Symbol("java.lang.Float"), (jfloat,), Float32(n))
+convert(::Type{JavaObject{Symbol("java.lang.Long")}}, n::Real) = jnew(Symbol("java.lang.Long"), (jlong,), Int64(n))
+convert(::Type{JavaObject{Symbol("java.lang.Integer")}}, n::Real) = jnew(Symbol("java.lang.Integer"), (jint,), Int32(n))
+convert(::Type{JavaObject{Symbol("java.lang.Short")}}, n::Real) = jnew(Symbol("java.lang.Short"), (jshort,), Int16(n))
+convert(::Type{JavaObject{Symbol("java.lang.Byte")}}, n::Real) = jnew(Symbol("java.lang.Byte"), (jbyte,), Int8(n))
+convert(::Type{JavaObject{Symbol("java.lang.Character")}}, n::Real) = jnew(Symbol("java.lang.Character"), (jchar,), Char(n))
+convert(::Type{JavaObject{:int}}, n) = convert(jint, n)
+convert(::Type{JavaObject{:long}}, n) = convert(jlong, n)
+convert(::Type{JavaObject{:byte}}, n) = convert(jbyte, n)
+convert(::Type{JavaObject{:boolean}}, n) = convert(jboolean, n)
+convert(::Type{JavaObject{:char}}, n) = convert(jchar, n)
+convert(::Type{JavaObject{:short}}, n) = convert(jshort, n)
+convert(::Type{JavaObject{:float}}, n) = convert(jfloat, n)
+convert(::Type{JavaObject{:double}}, n) = convert(jdouble, n)
+convert(::Type{JavaObject{:void}}, n) = convert(jvoid, n)
+convert(::Type{JavaObject{T}}, ::Nothing) where T = jnull
```
These convert methods seem fine to merge.

mkitti commented on src/convert.jl:
```diff
@@ -1,19 +1,34 @@
+    if isnull(obj)
+        return JavaObject{T}(obj.ptr)
+    elseif isConvertible(T, S)   #Safe static cast
+        return JavaObject{T}(obj.ptr)
     end
-    isnull(obj) && throw(ArgumentError("Cannot convert NULL"))
     realClass = ccall(jnifunc.GetObjectClass, Ptr{Nothing}, (Ptr{JNIEnv}, Ptr{Nothing} ), penv, obj.ptr)
     if isConvertible(T, realClass)  #dynamic cast
-        ptr = ccall(jnifunc.NewLocalRef, Ptr{Nothing}, (Ptr{JNIEnv}, Ptr{Nothing}), penv, obj.ptr)
-        ptr === C_NULL && geterror()
-        return JavaObject{T}(ptr)
+        return JavaObject{T}(obj.ptr)
     end
     throw(JavaCallError("Cannot cast java object from $S to $T"))
 end
```
This allows for null objects. We could probably merge this.

mkitti commented on src/core.jl:
```diff
@@ -82,16 +132,22 @@ isnull(obj::JavaMetaClass) = obj.ptr == C_NULL
 const JString = JavaObject{Symbol("java.lang.String")}
+const jnull = JavaObject(nothing)
+
+JavaObject(ptr::Ptr{Nothing}) = ptr == C_NULL ? JavaObject(ptr) : JavaObject{Symbol(getclassname(getclass(ptr)))}(ptr)
```
This depends on functions in proxy.jl at the moment, so we will probably move it into proxy.jl

zot commented:
Folks, do you need any more information from me in order to tie this off?

mkitti commented:
Hi @zot, thanks for checking in. JavaCall has undergone some major structural changes since a few weeks ago, so this will need to be updated again.

Next on my list is to improve JavaCall's local/global reference management. However, I think the reorganization will make things much more manageable moving forward. For example, all the `ccall` should be wrapped into the new `JNI` submodule.

Do we have your permission to continue developing this? My plan remains making this a separate package which has JavaCall as a dependency.

zot commented:
> Do we have your permission to continue developing this? My plan remains making this a separate package which has JavaCall as a dependency.

Of course! JavaCall belongs to you guys -- I wrote the proxy code in order to contribute to the project. Use and modify it as you see fit, I'd just like to see it merged in one form or another...

mkitti commented:
@zot , @aviks I split the proxy code off into its own `JProxies` subpackage.

1. I switched most of the JNI calls over to the `JavaCall.JNI` module which wraps the `ccalls`
2. I disabled most of the memory management code since I want to use `JNI.PushLocalFrame` and `JNI.PopLocalFrame` in the future which already does the job of tracking local references and marking them for garbage collection.
3. All tests in both packages are passing.

The easiest way I found to load `JProxies` is to symlink the JProxies folder into ~/.julia/dev and then use `Pkg.develop`. More work is required, but I think this puts us on a more sustainable path.

zot commented on JProxies/doc/index.md:
```diff
@@ -0,0 +1,44 @@
+* Create static proxies to access static members
+Primitive types and strings are converted to Julia objects on field accesses and method returns and converted back to Java types when sent as arguments to Java methods.
+*NOTE: Because of this, if you need to call Java methods on a string that you got from Java, you'll have to use `JProxy(str)` to convert the Julia string to a proxied Java string*
+To invoke static methods, set static to true (see below).
```
I wrote this before I added the `@class` macro -- I think it should change to this:

`To invoke static methods, use @class (see below).`

zot commented on JProxies/doc/index.md:
```diff
@@ -0,0 +1,44 @@
+julia> JProxy(@jimport(java.lang.System)).getName()
+"java.lang.System"
+
+julia> JProxy(@jimport(java.lang.System);static=true).out.println("hello")
```
I'd change this to
`julia> @class(java.lang.System).out.println("hello")`

aviks commented:
Can we just make this a submodule for the JavaCall module, so doing `using JavaCall; using JavaCall.JProxy` gets you the new behaviour?

mkitti commented:
> Can we just make this a submodule for the JavaCall module, so doing `using JavaCall; using JavaCall.JProxy` gets you the new behaviour?

We could do that although the direction I was heading in was to spin it out so that it can live as its own package. Currently, you could import `JProxies` without explicitly importing `JavaCall`.

```julia
using JProxies
JProxies.init()
```

Under the hood that will still use JavaCall. You made a point once that there was a lot of the proxy code to support which could make future maintenance more difficult. Dividing the code clearly like this is beneficial for maintainability I think.

There is still some code I actually want to move into JavaCall itself. For example, `listfields` probably belongs in reflect.jl.

zot commented:
> > Can we just make this a submodule for the JavaCall module, so doing `using JavaCall; using JavaCall.JProxy` gets you the new behaviour?
> 
> We could do that although the direction I was heading in was to spin it out so that it can live as its own package. Currently, you could import `JProxies` without explicitly importing `JavaCall`.
> 
> ```julia
> using JProxies
> JProxies.init()
> ```
> 
> Under the hood that will still use JavaCall. You made a point once that there was a lot of the proxy code to support which could make future maintenance more difficult. Dividing the code clearly like this is beneficial for maintainability I think.

That sounds like a good direction to me. That way people who just want to use JavaCall don't need to load in JProxy and people who want JProxy don't need to worry about JavaCall.

> There is still some code I actually want to move into JavaCall itself. For example, `listfields` probably belongs in reflect.jl.

Yes, I wrote some useful utility code that could move into JavaCall but I wanted to keep my changes to the base package to a minimum so I kept as much as I could in proxy.jl.

zot commented:
Is this dying on the vine?

mkitti commented:
Most of the changes to core JavaCall were merged in b5692cc14d7830283b90383a3b15fe752518c47e a few months ago.

We're preparing for a JavaCall 0.8 release. Taro just set compatibility with 0.8.

It looks like we have conflicts to fix here. The `jnull` thing should be tested again.

mkitti commented:
Add a JProxy subpackage
